metadata/src/main/java/org/apache/kafka/image/AclsDelta.java

```
@@ -61,7 +61,13 @@ public void replay(AccessControlEntryRecord record) {
     }
 
     public void replay(RemoveAccessControlEntryRecord record) {
-        changes.put(record.id(), Optional.empty());
+        if (image.acls().containsKey(record.id())) {
+            changes.put(record.id(), Optional.empty());
+        } else if (changes.containsKey(record.id())) {
+            changes.remove(record.id());
+        } else {
+            throw new RuntimeException("Failed to find existing ACL with ID " + record.id() + " in either image or changes");
```

Can we throw an `IllegalStateException`?

---

I agree that exception is more appropriate. I went with `RuntimeException` because thats what we throw in `AclControlManager` (https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/controller/AclControlManager.java#L202) when we come across a removal unexpectedly. My thought process was to be consistent. Having said that, they are two different components so I'm fine updating it here to `IllegalStateException`. Do you think we should throw the same exception in `AclControlManager`?